src/shacl/ShaclParser.ts

```diff
@@ -0,0 +1,583 @@
+      semanticErrors,
+      cst,
+    };
+  };
+
+  constructor(
+    config?: Partial<IParserConfig>,
+    prefixes = { shacl: 'sh', xsd: 'xsd' }
+  ) {
+    super(
+      {
+        outputCst: true,
+        recoveryEnabled: true,
+        ...config,
+      },
+      getShaclTokenTypes(prefixes),
+      getShaclTokenTypes(prefixes),
+      false
+    );
+
+    this.prefixes = prefixes;
+    this.lexer = new Lexer(getShaclTokenTypes(prefixes));
+    this.shaclTokenMap = getShaclTokenMap(prefixes);
+    Parser.performSelfAnalysis(this);
+  }
+
+  predicateObjectList = this.OVERRIDE_RULE('predicateObjectList', () => {
+    this.OR([
+      {
+        ALT: () => {
+          this.SUBRULE(this.shaclRulePredicateObjectList);
+        },
+      },
+      {
+        ALT: () => {
+          this.SUBRULE(this.verb);
+          this.SUBRULE(this.objectList);
+        },
+      },
+    ]);
+    this.MANY(() => {
+      this.CONSUME(turtleTokenMap.Semicolon);
+      this.OPTION(() => {
+        this.OR1([
+          {
+            ALT: () => {
+              this.SUBRULE1(this.shaclRulePredicateObjectList);
+            },
+          },
+          {
+            ALT: () => {
+              this.SUBRULE1(this.verb);
+              this.SUBRULE1(this.objectList);
+            },
+          },
+        ]);
+      });
+    });
+  });
+
+  shaclRulePredicateObjectList = this.RULE(
+    'shaclRulePredicateObjectList',
+    () => {
+      this.OR([
+        {
+          ALT: () => this.SUBRULE(this.shaclPredicateIRI),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclNodeKind),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclTargetNode),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclPropertyPath),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclLiteralConstraint),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclListTakingConstraint),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclShapeExpectingConstraint),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclHasValueConstraint),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclVerbShape),
+        },
+      ]);
+    }
+  );
+
+  shaclPredicateIRI = this.RULE('shaclPredicateIRI', () => {
+    this.CONSUME(categoryTokenMap.IriTakingPredicate);
+    this.SUBRULE(this.iri); // FIXME? This may need to allow a list of IRIs, as with shaclTargetNode.
+  });
+
+  shaclNodeKind = this.RULE('shaclNodeKind', () => {
+    this.CONSUME(this.shaclTokenMap.SHACL_nodeKind);
+    this.CONSUME(categoryTokenMap.NodeKindIRI);
+  });
+
+  shaclTargetNode = this.RULE('shaclTargetNode', () => {
+    this.CONSUME(this.shaclTokenMap.SHACL_targetNode);
+    this.SUBRULE(this.shaclIRIOrLiteral);
+    this.MANY(() => {
+      this.CONSUME(turtleTokenMap.Comma);
+      this.SUBRULE1(this.shaclIRIOrLiteral);
+    });
+  });
+
+  shaclVerbShape = this.RULE('shaclVerbShape', () => {
```

Comment:
Does this mean there's either a verb (either `a` or an actual predicate) followed by a shape type (`sh:Shape`, `sh:NodeShape`, or `sh:PropertyShape`) optionally followed by a comma-separated list of more shape types?

If so, is `this.verb` too broad here or am I following the code wrong? (should it only allow `a` or `rdf:type`?)

Reply:
I'm not sure. What you're discussing _is_ the primary intent, but using `this.verb` prevents the parser from having issues parsing stuff like the following:

```
ex:SomeThing a sh:NodeShape ;
  ex:relatesTo ex:SomeOtherThing .
```

I'm not totally sure there's any great reason to arbitrarily embed non-SHACL-style Turtle but I _think_ it is probably allowed (though I'm not totally sure).

Reply:
Would it be arbitrary non-SHACL-style Turtle or just arbitrary predicates?